src/knapsack/BaseKnapsackModel.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2017 Battelle Energy Alliance, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Created on February 2, 2021
+
+@author: mandd
+"""
+
+#External Modules---------------------------------------------------------------
+import numpy as np
+import math
+import copy
+#External Modules End-----------------------------------------------------------
+
+#Internal Modules---------------------------------------------------------------
+from PluginsBaseClasses.ExternalModelPluginBase import ExternalModelPluginBase
+#Internal Modules End-----------------------------------------------------------
+
+
+class BaseKnapsackModel(ExternalModelPluginBase):
+  """
+    This class is designed to create the BaseKnapsack model
+  """
+  def __init__(self):
+    """
+      Constructor
+      @ In, None
+      @ Out, None
+    """
+    ExternalModelPluginBase.__init__(self)
+    
+    self.capacity      = None    # capacity value of the knapsack
+    self.penaltyFactor = 1.0     # penalty factor that is used when the capacity constraint is not satisfied
+    self.outcome       = None    # ID of the variable which indicates if the chosen elements satisfy the capacity constraint
+    self.choiceValue   = None    # ID of the variable which indicates the sum of the values of the chosen project elements
+    
+  def _readMoreXML(self, container, xmlNode):
+    """
+      Method to read the portion of the XML that belongs to the BaseKnapsack model
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, xmlNode, xml.etree.ElementTree.Element, XML node that needs to be read
+      @ Out, None
+    """
+    container.mapping    = {}
+    
+    for child in xmlNode:
+      if child.tag == 'capacity':
+        self.capacity = float(child.text.strip())
+      elif child.tag == 'penaltyFactor':
+        self.penaltyFactor = float(child.text.strip())
+      elif child.tag == 'outcome':
+        self.outcome = child.text.strip()
+      elif child.tag == 'choiceValue':
+        self.choiceValue = child.text.strip()
+      elif child.tag == 'map':
+        container.mapping[child.text.strip()] = [child.get('value'),child.get('cost')]
+      elif child.tag == 'variables':
+        variables = [str(var.strip()) for var in child.text.split(",")]
```

Comment:
This is the External Model for RAVEN, I think we'd better to use InputData and InputTypes to process the xml input.

Reply:
I have checked other External Models inside LOGOS, they are not using the InputData yet. If you want to use InputData for LOGOS, I suggest we open an issue, and adapt the InputData in another PR.

Reply:
Issue #9 has been opened.
We should have a chat on how to fix it.